### PR TITLE
Release/stable

### PR DESCRIPTION
**Pull request by Sceat**

### CODE REVIEW

File: example/client/simple.js [hunk 1]
```diff
@@ -1,41 +1,45 @@
...
-//     onMessage
-//   }
+import casual from 'casual'
+import {
+  inspect,
+} from 'util'
+
+import {
+  Node_Client,
+} from '../../src/index.js'
```
Comment (Sceat) [inline example/client/simple.js:9]: will need to pop out a web_client before releasing

File: package.json [hunk 2]
```diff
@@ -1,14 +1,17 @@
 {
   "name": "@hydre/graphql-websocket",
   "version": "1.0.6",
-  "description": "A Websocket GraphQL server and client",
+  "description": "A blazing fast graphql websocket server and client",
```
Comment (Sceat) [inline package.json:4]: keywords-sama :point_up:

File: src/server.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+          headers,
+        }, contextValue,
+    ) => {
+      const client_id = headers['sec-websocket-key']
+      const log_peer = log.extend(client_id)
+
+      log_peer('connected!')
+
+      const executor = new Executor({
+        id             : client_id,
```
Comment (Sceat) [inline src/server.js:107]: should add a comment here to inform that the id is just debug stuff and actually even remove it as it already is scoped under a ws

File: example/client/simple.js (hunk 1 above)
Comment (Sceat) [reply example/client/simple.js]: Done

File: src/client.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+} from 'util'
+import WebSocket from 'ws'
+import invariant from 'invariant'
+
+const log = debug('gql-ws').extend('client')
+const {
+  stripIgnoredCharacters, parse,
+} = graphql
+const DEFAULT_TIMEOUT = 31_000
+const util_promisify =
```
Comment (DeltaEvo) [inline src/client.js:18]: Why when you have `promisify` from util ?

File: src/server.js (hunk 3 above)
Comment (Sceat) [reply src/server.js:107]: still usefull to leg per client execution, may not impact anything actually

File: src/client.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        console.error(error)
+      }
+    })
+    log('disconnected')
+  }
+
+  heartbeat() {
+    clearTimeout(this.#ws.timeout)
+    this.#ws.timeout = setTimeout(() => {
+      this.#ws.terminate()
```
Comment (DeltaEvo) [inline src/client.js:90]: terminate ?

File: src/server.js [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    timeout = DEFAULT_TIMEOUT,
+  }) {
+    this.#ws_options = ws_options
+    this.#schema = schema
+    this.#rootValue = rootValue
+    this.#web_server = web_server
+    this.#context = context
+    this.#timeout = timeout
+
+    this.#context_is_static = typeof context === 'object'
```
Comment (DeltaEvo) [inline src/server.js:62]: Do you realy need a variable for this ?

File: src/server.js [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    const middleware = compose(this.#middleware)
+    const http_server = this.#web_server.listen(...parameters)
+    // we prevent usage of those as it's up to the http server to decide
+    // @see https://github.com/websockets/ws/blob/master/doc/ws.md#new-websocketserveroptions-callback
+    const {
+      // throwing off both vars
+      // eslint-disable-next-line no-unused-vars
+      host,
+      // eslint-disable-next-line no-unused-vars
+      port,
```
Comment (DeltaEvo) [inline src/server.js:88]: I think i migth be better to add a `varsIgnorePattern` for the eslint rule like if you variable start with `_`

File: src/server.js [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    })
+
+    const interval = setInterval(() => {
+      wss.clients.forEach(ws => {
+        if (!ws.alive) {
+          ws.force_terminate()
+          return
+        }
+
+        ws.alive = false
```
Comment (DeltaEvo) [inline src/server.js:218]: You realy need to handle this yourself ?

File: src/stream.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+export default async () => {
+  // eslint-disable-next-line no-undef
+  const stream = await globalThis.process?.release?.name ?
+    import('stream') : import('readable-stream')
```
Comment (DeltaEvo) [inline src/stream.js:4]: Do you really need this, webpack don't do this job already ?

File: src/client.js (hunk 4 above)
Comment (Sceat) [reply src/client.js:18]: in the browser ?

File: src/client.js (hunk 5 above)
Comment (Sceat) [reply src/client.js:90]: https://github.com/websockets/ws/blob/master/doc/ws.md#websocketterminate

File: src/server.js (hunk 7 above)
Comment (Sceat) [reply src/server.js:88]: event better i found there is a `ignoreRestSiblings`, added in [`24d92`](https://github.com/usidy/eslint-config-sidy/commit/24d9209ab5f1e14488183089a1595c276142d5f5)

File: src/server.js (hunk 8 above)
Comment (Sceat) [reply src/server.js:218]: seems so https://github.com/websockets/ws#how-to-detect-and-close-broken-connections

File: src/stream.js (hunk 9 above)
Comment (Sceat) [reply src/stream.js:4]: hum actually i have no idea, i saw browserify use another package, and i don't assume that this client is webpack-dependant :thinking:

File: src/server.js (hunk 6 above)
Comment (Sceat) [reply src/server.js:62]: indeed not

File: src/client.js (hunk 4 above)
Comment (DeltaEvo) [reply src/client.js:18]: Webpack will polyfill it

File: src/server.js [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+              // this could be an unsubscribe request from the client
+              // or a simple mistake.
+              // in any case we don't want duplicated operations
+              if (user_streams.has(operation_id)) {
+                const stream = user_streams.get(operation_id)
+                log_peer('operation %O was terminated', operation_id)
+                stream.end()
+
+                continue
+              }
```
Comment (Sceat) [inline src/server.js:143]: this is more elegant and simple than before

File: src/client.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
+import debug from 'debug'
+import graphql from 'graphql'
+
+const log = debug('gql-ws').extend('client')
+const {
+  stripIgnoredCharacters, parse,
+} = graphql
+const NORMAL_CLOSURE = 1000
+
+export default (PassThrough, WebSocket) =>
```
Comment (DeltaEvo) [inline src/client.js:10]: Nice

File: src/client.js [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+      this.ws.close(NORMAL_CLOSURE, 'closed by client')
+      log('disconnected')
+    }
+
+    async query(query, variables = {}) {
+      if (typeof query !== 'string')
+        throw new Error('The query is not a String')
+
+      // fail fast
+      parse(query)
```
Comment (DeltaEvo) [inline src/client.js:55]: You really need to import the whole graphql parser just for this ?

File: src/client.js [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+            id: operation_id,
+          }))
+          pass_through.end()
+          log('operation %O terminated', operation_id)
+        },
+        /**
+         * allows for...of
+         */
+        async *[Symbol.asyncIterator]() {
+          yield* pass_through
```
Comment (DeltaEvo) [inline src/client.js:120]: I think you can avoid the pass_trough by with something like this https://github.com/rolftimmermans/event-iterator

File: src/client.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+        data,
+      }) => {
+        const operation_response = JSON.parse(data.toString())
+        const {
+          id, end,
+        } = operation_response
+        if (id !== operation_id) return
+        if (end) pass_through.end()
+        else pass_through.write(operation_response)
+      }
```
Comment (DeltaEvo) [inline src/client.js:80]: Pushing a lot of message_handle that do all the json parsing seems to not be a good idea I think you should create an [EventTarget](https://developer.mozilla.org/fr/docs/Web/API/EventTarget) that dispatch event based on the operation_id

File: README.md [hunk 15]
```diff
@@ -1,31 +1,80 @@
...
+```
+
 ## Quick start
 
-## Documentation
\ No newline at end of file
+### Server
+```js
+import Server from '@hydre/graphql-websocket/server'
+// coming soon
```
[the remark this answers is not in the document]
Comment (Sceat) [reply README.md:40]: basically that mean before 2025

File: src/client.js (hunk 12 above)
Comment (Sceat) [reply src/client.js]: removed

File: src/client.js (hunk 13 above)
Comment (Sceat) [reply src/client.js]: done :heavy_check_mark:

File: src/client.js (hunk 14 above)
Comment (Sceat) [reply src/client.js]: now using EventTarget :heavy_check_mark:

**Pull request merged**